trie/triedb/pathdb/journal.go

```diff
@@ -260,9 +260,37 @@ func (db *Database) loadLayers() layer {
 	if !(root == types.EmptyRootHash && errors.Is(err, errMissJournal)) {
 		log.Info("Failed to load journal, discard it", "err", err)
 	}
-	// Return single layer with persistent state.
-	nb := NewTrieNodeBuffer(db.diskdb, db.config.TrieNodeBufferType, db.bufferSize, nil, 0, db.config.ProposeBlockInterval, db.config.NotifyKeep)
-	dl := newDiskLayer(root, rawdb.ReadPersistentStateID(db.diskdb), db, nil, nb)
+
+	var (
+		nb      trienodebuffer
+		dl      *diskLayer
+		stateID = rawdb.ReadPersistentStateID(db.diskdb)
+	)
+	if errors.Is(err, errMissJournal) && db.config.EnableRecoverNodeBufferList && db.config.TrieNodeBufferType == NodeBufferList {
+		if db.freezer == nil {
+			log.Crit("Use unopened freezer db to recover node buffer list")
+		}
+		log.Info("Recover node buffer list from ancient db")
+		nb, err = NewTrieNodeBuffer(db.diskdb, db.config.TrieNodeBufferType, db.bufferSize, nil, 0,
+			db.config.ProposeBlockInterval, db.config.NotifyKeep, db.freezer, true)
+		if err != nil {
+			log.Error("Failed to new trie node buffer for recovery", "error", err)
+		} else {
+			root, stateID, _ = nb.getLatestStatus()
+			log.Info("Finish recovering node buffer list", "latest root hash", root.String(), "latest state_id", stateID)
```

Comment:
Add the total replay time cost to the log.

Reply:
Added